### PR TITLE
Added optimizer support regarding hash partitions.

### PR DESCRIPTION
**Pull request by zuyu**
This PR added optimizer support for hash partitions in CreateTable / Insert / InsertSelection / UpdateTable.

Re-assigned to @jianqiao. Thanks!

### CODE REVIEW

**Comment by hbdeshmukh** [comment]
Hi @zuyu, Would you mind assigning @jianqiao to review this PR? I believe he will be a better reviewer for this PR.

File: query_optimizer/resolver/Resolver.cpp [hunk 1]
```diff
@@ -478,9 +484,32 @@ L::LogicalPtr Resolver::resolveCreateTable(
...
-  return L::CreateTable::Create(relation_name, attributes, block_properties);
+  std::shared_ptr<const S::PartitionSchemeHeader> partition_scheme_header_proto(
+      resolvePartitionClause(create_table_statement));
+
+  return L::CreateTable::Create(relation_name, attributes, block_properties, partition_scheme_header_proto);
+}
+
+namespace {
+
+attribute_id GetAttributeIdFromName(const PtrList<ParseAttributeDefinition>&attribute_definition_list,
```
Comment (jianqiao) [inline query_optimizer/resolver/Resolver.cpp]: Need a space between type and parameter here.

File: query_optimizer/logical/CreateTable.hpp [hunk 2]
```diff
@@ -100,8 +108,9 @@ class CreateTable : public Logical {
   static CreateTablePtr Create(
       const std::string &relation_name,
       const std::vector<expressions::AttributeReferencePtr> &attributes,
-      const std::shared_ptr<const StorageBlockLayoutDescription> &block_properties) {
-    return CreateTablePtr(new CreateTable(relation_name, attributes, block_properties));
+      const std::shared_ptr<const StorageBlockLayoutDescription> &block_properties,
+      const std::shared_ptr<const serialization::PartitionSchemeHeader> &partition_scheme_header_proto) {
```
Comment (jianqiao) [inline query_optimizer/logical/CreateTable.hpp]: Add a comment for `partition_scheme_header_proto` in method header (and mention that `partition_scheme_header_proto` can be `nullptr` if not specified).

**Comment by jianqiao** [comment]
LGTM except some minor style issues.

File: query_optimizer/physical/CreateTable.hpp [hunk 3]
```diff
@@ -107,8 +115,9 @@ class CreateTable : public Physical {
   static CreateTablePtr Create(
       const std::string &relation_name,
       const std::vector<expressions::AttributeReferencePtr> &attributes,
-      const std::shared_ptr<const StorageBlockLayoutDescription> &block_properties) {
-    return CreateTablePtr(new CreateTable(relation_name, attributes, block_properties));
+      const std::shared_ptr<const StorageBlockLayoutDescription> &block_properties,
+      const std::shared_ptr<const serialization::PartitionSchemeHeader> &partition_scheme_header_proto) {
+    return CreateTablePtr(new CreateTable(relation_name, attributes, block_properties, partition_scheme_header_proto));
```
Comment (jianqiao) [inline query_optimizer/physical/CreateTable.hpp]: Add a comment for `partition_scheme_header_proto` in method header.

**Comment by zuyu** [comment]
Fixed the comments. Thanks!

**Pull request closed without merging**